### PR TITLE
feat(dart): add support pubspec.lock files

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
add support `pubspec.lock` files for `Dart`

### CODE REVIEW

File: pkg/dart/lock/parse.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+type Dep struct {
+	Dependency string `yaml:"dependency"`
+	Version    string `yaml:"version"`
+}
+
+func (Parser) Parse(r dio.ReadSeekerAt) ([]types.Library, []types.Dependency, error) {
+	l := &lock{}
+	decoder := yaml.NewDecoder(r)
+	if err := decoder.Decode(&l); err != nil {
+		return nil, nil, xerrors.Errorf("failed to decode pubspec.lock file: %s", err.Error())
```
Comment (knqyf263) [inline pkg/dart/lock/parse.go:37]:
```suggestion
		return nil, nil, xerrors.Errorf("failed to decode pubspec.lock file: %w", err)
```

Comment (knqyf263) [inline pkg/dart/lock/parse.go:37]:
```suggestion
		return nil, nil, xerrors.Errorf("failed to decode pubspec.lock file: %w", err)
```

File: pkg/dart/lock/parse.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+			// so find Indirect dependencies using "direct main" and "direct dev" fields
+			Indirect: dep.Dependency != directDevDep && dep.Dependency != directMainDep,
+		}
+		libs = append(libs, lib)
+	}
+
+	return libs, nil, nil
+}
+
+func parseDep(dep string) (types.Library, error) {
```
Comment (knqyf263) [inline pkg/dart/lock/parse.go:57]: What is this?

File: pkg/dart/lock/parse.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+func (Parser) Parse(r dio.ReadSeekerAt) ([]types.Library, []types.Dependency, error) {
+	l := &lock{}
+	decoder := yaml.NewDecoder(r)
+	if err := decoder.Decode(&l); err != nil {
+		return nil, nil, xerrors.Errorf("failed to decode pubspec.lock file: %s", err.Error())
+	}
+	var libs []types.Library
+	for name, dep := range l.Packages {
+		lib := types.Library{
+			ID:      fmt.Sprintf(idFormat, name, dep.Version),
```
Comment (knqyf263) [inline pkg/dart/lock/parse.go:42]: Shouldn't we use `pkgID()`?

File: pkg/dart/lock/parse.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+			Name:    name,
+			Version: dep.Version,
+			// there are 3 types of dependency fields:
+			// "direct main", "direct dev" and transitive
+			// transitive non-string field
```
Comment (knqyf263) [inline pkg/dart/lock/parse.go:47]: What do you mean? Isn't it string?
https://github.com/aquasecurity/go-dep-parser/blob/b4aa3d0aeaba9085595c32c48fdd3ae1d5ca3584/pkg/dart/lock/testdata/happy.lock#L17

**Review by knqyf263 — COMMENTED** [review]
I believe you mean `Dart`, not `dark`.

File: pkg/dart/lock/parse.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/dart/lock/parse.go]: yes. Thanks you. Fixed.

File: pkg/dart/lock/parse.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/dart/lock/parse.go]: Done

File: pkg/dart/lock/parse.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/dart/lock/parse.go]: This is my mistake. Sorry. Fixed.

**Comment by DmitriyLewen** [comment]
> I believe you mean Dart, not dark.

yes. It was typo.

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**